### PR TITLE
TAI Enhancements

### PR DESCRIPTION
**Pull request by scott-emery**
This is a fairly large modification to the TAI interface. It include a number
of modifications:

1. There are 3 APIs, and objects:
   a. Module - The optical module itself
   b. Network Interface - The optical interfaces leaving the module
   c. Host Interface - The internal interfaces connected to switching silicon
2. Module presence/absence indications from adapter to adapter host
3. Independence from SAI. All identifiers are changed from "SAI" to "TAI"
4. Documentation
   a. README.md describes the TAI interfaces/objects/attributes and their usage
   b. DOXYGEN comments for all APIs/objects/attributes
5. Removed SAI-specific functions which don't apply to TAI, like profiles.
6. Incorporate comments from pull request #1
   a. Included NTT's copyright on each file
   b. Added ifdef for proper C++ compilation
   c. Corrected misuse to "adapter" vs. "adapter host" in certain places.
   d. Removed all custom attributes. They can be added later.

### CODE REVIEW

**Comment by ishidawataru** [comment]
Could you also edit `stub_tai.c` and make it compatible with your change?

https://github.com/Telecominfraproject/OOPT-DTC-TAI/blob/master/stub/stub_tai.c

**Comment by scott-emery** [comment]
I was afraid you were going to ask for the stub_tai.c modifications too. I'll add that to this pull request in the next couple of days.

File: README.md [hunk 1]
```diff
@@ -1,2 +1,289 @@
...
+* attr_count and attr_list. A list of attributes which will be set upon the 
+module object when it is created. This can be used to override the default 
+attribute values for a module.
+* notification functions. A pointer to a list of function pointers in the 
+adapter host which the adapter can call to notify the adapter host of changes in 
+status of the module.
+
+The attr_list supplied on the `create_module()` function call must contain a 
+`TAI_MODULE_ATTR_LOCATION` attribute. This attribute defines which module is 
+being created. This is typically set the to value of the module_location
```
Comment (ishidawataru) [inline README.md]: set the to -> set to the

Comment (scott-emery) [reply README.md]: Ack.

**Comment by ishidawataru** [comment]
Thanks!

**Pull request merged**